File: 0733-flood-fill/0733-flood-fill.py
from collections import deque

class Solution:
    def floodFill(self, image: List[List[int]], sr: int, sc: int, color: int) -> List[List[int]]:
        ROWS = len(image)
        COLS = len(image[0])
        DIRECTIONS = [[0, 1], [0, -1], [1, 0], [-1, 0]]
        
        original_color = image[sr][sc]
        if original_color == color:
            return image

        image[sr][sc] = color
        queue = deque([(sr, sc)])

        while queue:
            r, c = queue.popleft()

            for dr, dc, in DIRECTIONS:
                nr = dr + r
                nc = dc + c

                if nr < 0 or nc < 0 or nr == ROWS or nc == COLS or image[nr][nc] != original_color:
                    continue
                
                image[nr][nc] = color
                queue.append((nr, nc))
        
        return image

        

File: 0733-flood-fill/0733-flood-fill.js
/**
 * @param {number[][]} image
 * @param {number} sr
 * @param {number} sc
 * @param {number} newColor
 * @return {number[][]}
 */
var floodFill = function(image, sr, sc, newColor) {
    const ROWS = image.length, COLS = image[0].length;
    let visited = new Set();
    let startingPixel = image[sr][sc];
    let directions = [[-1, 0], [0, -1], [1, 0], [0, 1]];
    
    dfs( [ [sr, sc] ]);
    return image;
    
    function dfs( queue ) {
        while( queue.length ) {
            let [ r, c ] = queue.pop();
            image[r][c] = newColor;
            
            for( let direction of directions ) {
                let [ rD, cD ] = direction;
                rD += r;
                cD += c;
                
                if( rD >= 0 && cD >= 0 && rD < ROWS && cD < COLS && image[rD][cD] === startingPixel && !visited.has(`${rD}-${cD}`) ) {
                    queue.push([rD, cD]);
                    visited.add(`${rD}-${cD}`);
                }
            }
        }
    }
};